### PR TITLE
Support reading reference sequence from Hadoop

### PR DESCRIPTION
**Pull request by tomwhite**
All ways of reading reference sequences are hidden behind a common interface,
ReferenceDataflowSource. See https://github.com/broadinstitute/hellbender/pull/804 for initial discussion of the design, and https://github.com/broadinstitute/hellbender/issues/567 for the issue that this solves.

This change adds some HTSJDK extensions to work with Hadoop, these should eventually be added to HTSJDK in some form.

I was able to run BQSR on a Spark cluster using this code, by using the following:

```bash
# Copy BAM into HDFS
hadoop fs -put ./src/test/resources/org/broadinstitute/hellbender/tools/BQSR/NA12878.chr17_69k_70k.dictFix.bam bam/NA12878.chr17_69k_70k.dictFix.bam

# Copy FASTA (reference) into HDFS
hadoop fs -mkdir fasta 
hadoop fs -put ./src/test/resources/human_g1k_v37.chr17_1Mb.fasta fasta/human_g1k_v37.chr17_1Mb.fasta
hadoop fs -put ./src/test/resources/human_g1k_v37.chr17_1Mb.dict fasta/human_g1k_v37.chr17_1Mb.dict
hadoop fs -put ./src/test/resources/human_g1k_v37.chr17_1Mb.fasta.fai fasta/human_g1k_v37.chr17_1Mb.fasta.fai

spark-submit \
       --master $SPARK_MASTER \
       --conf spark.driver.userClassPathFirst=true \
       --conf spark.executor.userClassPathFirst=true \
       --conf spark.io.compression.codec=lzf \
       build/libs/hellbender-all-*-spark.jar ApplyWholeBQSRDataflow \
         --reference hdfs://$NAMENODE/user/$USER/fasta/human_g1k_v37.chr17_1Mb.fasta \
         --input hdfs://$NAMENODE/user/$USER/bam/NA12878.chr17_69k_70k.dictFix.bam \
         --output hdfs://$NAMENODE/user/$USER/out/bqsr \
         --knownSites $HELLBENDER_HOME/src/test/resources/org/broadinstitute/hellbender/tools/BQSR/dbsnp_132.b37.excluding_sites_after_129.chr17_69k_70k.vcf \
         --RECAL_TABLE_FILE /dev/stdout \
         --runner SPARK \
         --sparkMaster $SPARK_MASTER 
```

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ReferenceDataflowSource.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+     */
+    public ReferenceDataflowSource(final PipelineOptions pipelineOptions, final String reference,
+                                   final SerializableFunction<GATKRead, SimpleInterval> referenceWindowFunction) {
+        if (isFasta(reference)) {
+            if (BucketUtils.isHadoopUrl(reference)) {
+                referenceSource = new ReferenceHadoopSource(reference);
+            } else {
+                referenceSource = new ReferenceFileSource(reference);
+            }
+        } else { // use the Google Genomics API
```
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ReferenceDataflowSource.java]: Please make this an else if and throw in the new else.

Comment (tomwhite) [inline src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ReferenceDataflowSource.java]: Thanks for taking a look @davidadamsphd. What do you think the condition for this should be? The `reference` is an arbitrary ID (e.g. EOSt9JOVhp3jkwE), so there's no good way of testing for that case. We might check if the `pipelineOptions` can be converted to `GCSOptions` and that the necessary credentials are set, but that is done by `RefAPISource` anyway, which is the natural place to check it.

File: src/test/java/org/broadinstitute/hellbender/engine/dataflow/transforms/RefBasesFromAPIUnitTest.java [hunk 2]
```diff
@@ -43,11 +39,13 @@
 
 public final class RefBasesFromAPIUnitTest extends BaseTest {
 
-    private RefAPISource createMockRefAPISource( final List<SimpleInterval> intervals ) {
-        RefAPISource mockSource = mock(RefAPISource.class, withSettings().serializable());
+    private ReferenceDataflowSource createMockReferenceDataflowSource(final List<SimpleInterval> intervals,
```
Comment (davidadamsphd) [inline src/test/java/org/broadinstitute/hellbender/engine/dataflow/transforms/RefBasesFromAPIUnitTest.java]: Maybe this should be public and used in the other tests.

File: src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ReferenceDataflowSource.java (hunk 1 above)
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ReferenceDataflowSource.java]: I was initially thinking about checking for GCSOptions, but I agree it should happen in RefAPISource.
I just asked internally what the bounds are for the string. This is fine as-is until I hear back (or the answer is too complex).

File: build.gradle [hunk 3]
```diff
@@ -271,6 +271,7 @@ shadowJar {
     zip64 true
     dependencies {
         // exclude Hadoop dependencies, since they are provided when running with Spark
+        // TODO: improve this, since it doesn't take transitive dependencies into account
```
Comment (davidadamsphd) [inline build.gradle]: Optional: file a ticket for this to track it.

File: src/test/java/org/broadinstitute/hellbender/engine/dataflow/ReferenceHadoopSourceUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+public class ReferenceHadoopSourceUnitTest extends BaseTest implements Serializable {
+    private static final long serialVersionUID = 1L;
+
+    @Test
+    public void testReferenceSourceQuery() throws IOException {
+        final String referencePath = hg19MiniReference;
+
+        final ReferenceHadoopSource refSource = new ReferenceHadoopSource(referencePath);
+        final ReferenceBases bases = refSource.getReferenceBases(PipelineOptionsFactory.create(),
```
Comment (davidadamsphd) [inline src/test/java/org/broadinstitute/hellbender/engine/dataflow/ReferenceHadoopSourceUnitTest.java]: Please add a comment explaining where hdfs comes in here. Looking at this, it looks like it's just reading a local file.

**Comment by davidadamsphd** [comment]
Looks good overall. A few general things:
(1) It's hard to diff the files copied from htsjdk and tell what's new (I don't want to add comments on code that's old). Please add comment(s) explain what's actually new.
(2) We need a plan to get this code into htsjdk quickly. I'm concerned about this code getting out of sync with htsjdk. I created a ticket #831 about this and we'll work with @droazen to solve it.

File: src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ReferenceDataflowSource.java (hunk 1 above)
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ReferenceDataflowSource.java]: So, there are limits, but nothing we should depend on. Leave this as-as.

File: src/test/java/org/broadinstitute/hellbender/engine/dataflow/transforms/RefBasesFromAPIUnitTest.java (hunk 2 above)
Comment (tomwhite) [inline src/test/java/org/broadinstitute/hellbender/engine/dataflow/transforms/RefBasesFromAPIUnitTest.java]: Let's leave this to a later refactoring if that's OK with you - this change is already pretty large.

File: build.gradle (hunk 3 above)
Comment (tomwhite) [inline build.gradle]: Opened https://github.com/broadinstitute/hellbender/issues/836 and added a reference to it in the comment.

File: src/test/java/org/broadinstitute/hellbender/engine/dataflow/ReferenceHadoopSourceUnitTest.java (hunk 4 above)
Comment (tomwhite) [inline src/test/java/org/broadinstitute/hellbender/engine/dataflow/ReferenceHadoopSourceUnitTest.java]: It is reading from a local file, but using the Hadoop logic. Added a comment.

**Comment by tomwhite** [comment]
I rebased on master and addressed the feedback. I've added comments (prefixed with "Hellbender") in the htsjdk code to make it very clear where the changes are.

I'll look at the changes needed in htsjdk next to address #831.

File: src/test/java/org/broadinstitute/hellbender/engine/dataflow/transforms/RefBasesFromAPIUnitTest.java (hunk 2 above)
Comment (davidadamsphd) [inline src/test/java/org/broadinstitute/hellbender/engine/dataflow/transforms/RefBasesFromAPIUnitTest.java]: sg.

**Comment by davidadamsphd** [comment]
It looks like you have some failing tests... I'm guessing JP's new BQSR. Otherwise, it looks ready to merge.

**Comment by tomwhite** [comment]
Rebased on master again.

**Comment by tomwhite** [comment]
Tests passed, this should be ready to merge now, @davidadamsphd.

**Comment by davidadamsphd** [comment]
LGTM. Merge, merge merge! :)

**Pull request merged**